### PR TITLE
CLDC-2140 Scheme and location updates

### PR DESCRIPTION
**Pull request by natdeanlewissoftwire**
this adds a few new features:

new complete/incomplete location breakdown: 
image

validation if a scheme has no completed locations: 
image

also, only active locations show on the locations question now and schemes and locations are only confirmed when they are complete and the user clicks "save"/"create" on the cya page

https://digital.dclg.gov.uk/jira/browse/CLDC-2140

### CODE REVIEW

File: app/models/scheme.rb [hunk 1]
```diff
@@ -166,7 +166,7 @@ def synonyms
   end
 
   def appended_text
-    "#{locations.count { |location| location.startdate.blank? || location.startdate <= Time.zone.today }} locations"
+    "#{confirmed_locations_count} completed #{'location'.pluralize(confirmed_locations_count)}, #{unconfirmed_locations_count} incomplete #{'location'.pluralize(unconfirmed_locations_count)}"
```
Comment (natdeanlewissoftwire) [inline app/models/scheme.rb]: we use language completed/incomplete on the frontend but actually this is just the same as confirmed/unconfirmed for purposes in the code. It made the most sense to juxtapose these here rather than earlier on e.g. by making a completed scope that was actually just checking for confirmed

File: spec/models/form/lettings/questions/location_id_spec.rb [hunk 2]
```diff
@@ -70,17 +72,19 @@
...
           FactoryBot.create(:location, scheme:, startdate: Time.utc(2023, 1, 1))
+          lettings_log.update!(scheme:)
         end
 
         it "the displayed_answer_options is an empty hash" do
           expect(question.displayed_answer_options(lettings_log)).to eq({})
         end
       end
 
-      context "and the locations have a no startdate" do
```
Comment (natdeanlewissoftwire) [inline spec/models/form/lettings/questions/location_id_spec.rb:80]: 🇮🇹

File: app/controllers/locations_controller.rb [hunk 3]
```diff
@@ -134,7 +134,9 @@ def update_availability
   def check_answers; end
 
   def confirm
-    flash[:notice] = "#{@location.postcode} #{@location.startdate.blank? || @location.startdate < Time.zone.now ? 'has been' : 'will be'} added to this scheme"
+    if @location.confirmed
+      flash[:notice] = "#{@location.postcode} #{@location.startdate.blank? || @location.startdate < Time.zone.now ? 'has been' : 'will be'} added to this scheme"
```
Comment (jamesrose) [inline app/controllers/locations_controller.rb]: This logic should be in a helper.

File: app/models/scheme.rb (hunk 1 above)
Comment (jamesrose) [reply app/models/scheme.rb]: This logic should not be in the model — look at moving to a helper.

File: app/models/scheme.rb [hunk 4]
```diff
@@ -264,4 +264,8 @@ def has_confirmed_locations?
   def confirmed_locations_count
     locations.confirmed.size
   end
+
+  def unconfirmed_locations_count
+    locations.unconfirmed.size
+  end
```
Comment (jamesrose) [inline app/models/scheme.rb]: You shouldn't need to extract these methods out.

File: app/models/validations/setup_validations.rb [hunk 5]
```diff
@@ -20,6 +20,14 @@ def validate_location(record)
     location_during_startdate_validation(record, :location_id)
   end
 
+  def validate_scheme_has_confirmed_locations_validation(record)
+    return unless record.scheme
+
+    unless record.scheme.has_confirmed_locations?
```
Comment (jamesrose) [inline app/models/validations/setup_validations.rb]:
```suggestion
    unless record.scheme.locations.confirmed.any?
```

File: app/models/validations/soft_validations.rb [hunk 6]
```diff
@@ -66,6 +66,12 @@ def all_tenants_age_and_gender_information_completed?
     end
   end
 
+  def all_tenants_gender_information_completed?
+    (1..hhmemb).all? do |n|
+      public_send("sex#{n}").present? && details_known_or_lead_tenant?(n)
+    end
+  end
```
Comment (jamesrose) [inline app/models/validations/soft_validations.rb]: Are these changes related to this PR?

File: app/controllers/locations_controller.rb [hunk 7]
```diff
@@ -134,7 +134,7 @@ def update_availability
   def check_answers; end
 
   def confirm
-    flash[:notice] = "#{@location.postcode} #{@location.startdate.blank? || @location.startdate < Time.zone.now ? 'has been' : 'will be'} added to this scheme"
+    flash[:notice] = helpers.location_creation_success_notice(@location)
```
Comment (natdeanlewissoftwire) [inline app/controllers/locations_controller.rb:137]: moved to helper

File: app/models/scheme.rb [hunk 8]
```diff
@@ -161,18 +161,6 @@ def check_support_attributes
...
-  end
-
-  def appended_text
-    "#{locations.count { |location| location.startdate.blank? || location.startdate <= Time.zone.today }} locations"
-  end
-
-  def hint
-    [primary_client_group, secondary_client_group].filter(&:present?).join(", ")
-  end
-
```
Comment (natdeanlewissoftwire) [inline app/models/scheme.rb:175]: moved to helper

File: app/models/validations/soft_validations.rb (hunk 6 above)
Comment (natdeanlewissoftwire) [reply app/models/validations/soft_validations.rb]: nope, just because diff is vs 1854 not main

File: app/views/form/_select_question.html.erb [hunk 9]
```diff
@@ -9,9 +9,9 @@
   hint: { text: question.hint_text&.html_safe }) do %>
   <% answers.each do |answer| %>
     <option value="<%= answer.id %>"
-      data-synonyms="<%= question.answer_option_synonyms(answer.resource) %>"
-      data-append="<%= question.answer_option_append(answer.resource) %>"
-      data-hint="<%= question.answer_option_hint(answer.resource) %>"
+      data-synonyms="<%= answer_option_synonyms(answer.resource) %>"
+      data-append="<%= answer_option_append(answer.resource) %>"
+      data-hint="<%= answer_option_hint(answer.resource) %>"
```
Comment (jamesrose) [inline app/views/form/_select_question.html.erb]: What can `answer.resource` be in all instances?

Comment (natdeanlewissoftwire) [reply app/views/form/_select_question.html.erb]: The resource is either "Select an option" or a scheme for the scheme typeahead, and for org typeaheads etc it's just the string of the org name so passing the resource allows us to test if it's an instance of a Scheme before we display Scheme specific synonyms/append/hint text

File: app/helpers/locations_helper.rb [hunk 10]
```diff
@@ -78,6 +78,12 @@ def toggle_location_link(location)
     return govuk_button_link_to "Reactivate this location", scheme_location_new_reactivation_path(location.scheme, location) if location.deactivated?
   end
 
+  def location_creation_success_notice(location)
+    if location.confirmed
+      "#{location.postcode} #{location.startdate.blank? || location.startdate < Time.zone.now ? 'has been' : 'will be'} added to this scheme"
+    end
```
Comment (arfacamble) [inline app/helpers/locations_helper.rb:84]: I believe the condition here can be slightly more concise `location.startdate&.before? Time.zone.now`

File: app/helpers/question_view_helper.rb [hunk 11]
```diff
@@ -13,6 +13,26 @@ def legend(question, page_header, conditional)
     }
   end
 
+  def answer_option_synonyms(resource)
+    return unless resource.instance_of?(Scheme)
```
Comment (arfacamble) [inline app/helpers/question_view_helper.rb:17]: I find `resource.is_a? Scheme` slightly more readable, but I'm pretty sure they are aliases, I jsut mention it in case you know of one and not the other

Comment (arfacamble) [inline app/helpers/question_view_helper.rb:17]: I find `resource.is_a? Scheme` slightly more readable, but I'm pretty sure they are aliases, I jsut mention it in case you know of one and not the other

File: app/helpers/question_view_helper.rb [hunk 12]
```diff
@@ -13,6 +13,26 @@ def legend(question, page_header, conditional)
...
+
+    confirmed_locations_count = resource.locations.confirmed.size
+    unconfirmed_locations_count = resource.locations.unconfirmed.size
+    "#{confirmed_locations_count} completed #{'location'.pluralize(confirmed_locations_count)}, #{unconfirmed_locations_count} incomplete #{'location'.pluralize(unconfirmed_locations_count)}"
+  end
+
+  def answer_option_hint(resource)
+    return unless resource.instance_of?(Scheme)
+
+    [resource.primary_client_group, resource.secondary_client_group].filter(&:present?).join(", ")
```
Comment (arfacamble) [inline app/helpers/question_view_helper.rb]: `.filter(&:present?)` does the same as `.compact` I believe.
Also, I may not have the full picture yet, but if both primary and secondary client groups are nil, then this will return an empty string. Is that fine?

**Review by arfacamble — CHANGES_REQUESTED** [review]
have got as far as _select_question.html.erb
note for me to remember, or kind of a question for you because I probably just don't understand. It feels weird to me that we have view helpers doing things that are model specific. Is there any way we can do this in a controller before we pass to a view?

File: spec/models/form/lettings/questions/location_id_spec.rb (hunk 2 above)
Comment (arfacamble) [reply spec/models/form/lettings/questions/location_id_spec.rb:80]: this is top quality bants,
turns out 🇱🇻 though

File: app/helpers/locations_helper.rb (hunk 10 above)
Comment (natdeanlewissoftwire) [reply app/helpers/locations_helper.rb:84]: will use the before but nil coalescing isn't equivalent here

File: app/helpers/question_view_helper.rb (hunk 11 above)
Comment (natdeanlewissoftwire) [reply app/helpers/question_view_helper.rb:17]: I prefer `instance_of?` in this case as it doesn't allow subclasses which I think is preferable conceptually

File: app/helpers/question_view_helper.rb (hunk 12 above)
Comment (natdeanlewissoftwire) [reply app/helpers/question_view_helper.rb]: only difference is .compact won't remove empty strings, but that's fine for this purpose - I'll change to .compact

File: app/helpers/locations_helper.rb (hunk 10 above)
Comment (arfacamble) [reply app/helpers/locations_helper.rb:84]: how comesies not equivalent?
location.startdate type is DateTime, so surely it's either a date or nil?

File: app/helpers/question_view_helper.rb (hunk 11 above)
Comment (arfacamble) [reply app/helpers/question_view_helper.rb:17]: rad

File: app/helpers/question_view_helper.rb (hunk 12 above)
Comment (arfacamble) [reply app/helpers/question_view_helper.rb]: wiki-wiki-wiki-wickeeeeed

File: app/helpers/locations_helper.rb (hunk 10 above)
Comment (natdeanlewissoftwire) [reply app/helpers/locations_helper.rb:84]: yep but we want nil to pass the ternary condition

Comment (arfacamble) [reply app/helpers/locations_helper.rb:84]: pleasure doing business sir

Comment (natdeanlewissoftwire) [reply app/helpers/locations_helper.rb:84]: 🤝

**Review by arfacamble — APPROVED** [review]
